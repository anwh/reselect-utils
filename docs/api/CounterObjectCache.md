# CounterObjectCache

## Description

[Re-reselect cache object](https://github.com/toomuchdesign/re-reselect/tree/master/src/cache) implementation with reference counter. When the selector instance has the number of references set to 0, the selector instance is automatically removed from the cache. This approach allows automatic cache memory management.

## Static Methods

### addRefRecursively(selector)(state, props)

#### Description

Recursively traverses the dependency chain of the selector and increments the number of references of each selector whose cache managed by CounterObjectCache.

#### Parameters

- _selector_ - base selector for recursive traverse.

- _state_ - state, which will passed to [selector.keySelector](https://github.com/toomuchdesign/re-reselect#selectorkeyselector).

- _props_ - props, which will passed to [selector.keySelector](https://github.com/toomuchdesign/re-reselect#selectorkeyselector).

#### Returns

Void.

### removeRefRecursively(selector)(state, props)

#### Description

Recursively traverses the dependency chain of the selector and decrements the number of references of each selector whose cache managed by CounterObjectCache.

#### Parameters

- _selector_ - base selector for recursive traverse.

- _state_ - state, which will passed to [selector.keySelector](https://github.com/toomuchdesign/re-reselect#selectorkeyselector).

- _props_ - props, which will passed to [selector.keySelector](https://github.com/toomuchdesign/re-reselect#selectorkeyselector).

#### Returns

Void.

## Constructor

#### Parameters

- _options_ - Configuration properties when instantiating a cache object. Properties:
  - _removeDelay_ - delay is before the instance of the selector is removed from the cache in milliseconds. Default value - 0.

## Instance Methods

### set(key, selector)

#### Parameters

- _key_ - key to add a selector instance to the cache.
- _selector_ - selector instance.

#### Returns

Void.

### get(key)

#### Parameters

- _key_ - to get the selector instance.

#### Returns

Selector instance.

### remove(key)

#### Parameters

- _key_ - key to remove the selector instance.

#### Returns

Void.

### clear()

#### Description

Clears the entire cache.

#### Returns

Void.

### isValidCacheKey(key)

#### Parameters

- _key_ - cache key to validate.

#### Returns

If the key is valid, it returns the `true`, otherwise returns `false`.

### addRef(key)

#### Parameters

- _key_ - cache key whose number of references needs to be incremented.

#### Returns

Void.

### removeRef(key)

#### Parameters

- _key_ - cache key whose number of references needs to be decremented.

#### Returns

Void.

## Examples
