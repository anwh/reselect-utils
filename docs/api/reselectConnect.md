# reselectConnect(selector, options)(Component)

## Description

A utility function that creates a special [
Higher-Order Component](https://reactjs.org/docs/higher-order-components.html). This is HOC that decorates the `Component` in such a way that the instances of `selector` and his dependencies will be created when `Component` is mounted, replaced when `Component` properties are updated, and when the `Component` is unmounted, the `selector` instance will be removed from cache.

## Parameters

- _selector_ - selector for synchronization.
- _options_ (optional) - Component sync settings object:
  - _updateRefsOnStateChange_ - flag that includes updating the refs and redrawing the component for each change of state. By default it has value `false`. Can be set to `true` if your selector or its dependencies use a state in a [keySelector](https://github.com/toomuchdesign/re-reselect#keyselector). **BE CAREFUL**: the activation of this flag can reduce the performance of your application, usually using the state in the `keySelector` is a very exotic case, try to avoid it.
- _Component_ - Component for decorating.

## Returns

New decorated Component whose life cycle is synchronized with the [CounterObjectCache](/docs/api/CounterObjectCache.md) instances of selector and his dependencies.

## Examples

```jsx
import createCachedSelector from 're-reselect';
import { CounterObjectCache, reselectConnect } from 'reselect-utils';

const personSelector = createCachedSelector(
  [state => state.persons, (state, props) => props.id],
  (persons, id) => ({
    person: persons[id],
  }),
)((state, props) => props.id, {
  cacheObject: new CounterObjectCache(),
});

const MyComponent = ({ person }) => <div>{person.name}</div>;

const ConnectedMyComponent = connect(personSelector)(
  // here used second decoration
  reselectConnect(personSelector)(MyComponent),
);
```
