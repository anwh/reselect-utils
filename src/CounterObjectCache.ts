import {ParametricSelector} from 'reselect';
import {ICacheObject} from 're-reselect';

type CacheItem = {
    selector: ParametricSelector<any, any, any>,
    refCount: number,
    removeTimeoutHandle?: number,
}

type Key = string | number;

type Cache = {
    [K in Key]: CacheItem
}

type CounterObjectCacheOptions = {
    removeDelay?: number
}

const defaultOptions: CounterObjectCacheOptions = {
    removeDelay: 0
};

export default class CounterObjectCache implements ICacheObject {
    private cache: Cache = {};
    private options: CounterObjectCacheOptions;

    constructor(options: CounterObjectCacheOptions = {}) {
        this.options = {
            ...defaultOptions,
            ...options
        }
    }

    set(key: Key, selector: ParametricSelector<any, any, any>) {
        this.cache[key] = {
            selector,
            refCount: 1,
        };
    }

    get(key: Key) {
        const cacheItem = this.cache[key];
        if (cacheItem === undefined) {
            return undefined
        }

        cacheItem.refCount++;
        clearTimeout(cacheItem.removeTimeoutHandle);

        return cacheItem.selector;
    }

    remove(key: Key) {
        const cacheItem = this.cache[key];
        cacheItem.refCount--;

        if (cacheItem.refCount === 0) {
            cacheItem.removeTimeoutHandle = setTimeout(() => {
                delete this.cache[key];
            }, this.options.removeDelay);
        }
    }

    clear() {
        Object.keys(this.cache)
            .forEach(key => this.remove(key));
    }

    isValidCacheKey(key: Key) {
        return typeof key === 'string' || typeof key === 'number';
    }
}