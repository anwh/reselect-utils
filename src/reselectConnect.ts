import {ComponentType, createElement, Component, ComponentClass} from "react";
import {connect} from "react-redux";
import {OutputParametricCachedSelector} from "re-reselect";
import {removeMatchingSelectorRecursively} from "./removeMatchingSelectorRecursively";
import {shallowEqual} from 'shallow-equal-object'

export const reselectConnect = <S, P, R>(
    selector: ReturnType<OutputParametricCachedSelector<S, P, R, any>>
) => <C extends ComponentType<P>>(WrappedComponent: C): ComponentClass<JSX.LibraryManagedAttributes<C, P>> => {

    const wrappedComponentName =
        WrappedComponent.displayName || WrappedComponent.name || 'Component';

    const clearCache = removeMatchingSelectorRecursively(selector);
    const state = Symbol('state');

    class Wrapper extends Component<P & { [state]: S }> {
        public static displayName = `ReselectConnect(${wrappedComponentName})`;

        public shouldComponentUpdate(nextProps: P & { [state]: S }) {
            return !shallowEqual(this.props, nextProps)
        }

        public componentDidUpdate(prevProps: P & { [state]: S }) {
            clearCache(prevProps[state], prevProps)
        }

        public componentWillUnmount() {
            clearCache(this.props[state], this.props)
        }

        public render() {
            return createElement(
                WrappedComponent,
                this.props
            )
        }
    }

    return connect(
        reduxState => ({[state]: reduxState}),
        {},
        undefined,
        {getDisplayName: name => `InnerConnect(${name})`}
        // @ts-ignore because something wrong with type Matching in react-redux typing
    )(Wrapper)
};