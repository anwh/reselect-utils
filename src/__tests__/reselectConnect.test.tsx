import * as React from "react";
import {FunctionComponent} from "react";
import {mount, shallow} from "enzyme";
import {compose, createStore} from "redux";
import {connect, Provider} from "react-redux";
import createCachedSelector from "re-reselect";
import CounterObjectCache from "../CounterObjectCache";
import {reselectConnect} from "../reselectConnect";

describe('reselectConnect', () => {
    type State = {
        value: string
    }

    type Props = {
        prop: string
    }

    test('should proxy all props and renders wrapped component', () => {
        const selector = createCachedSelector(
            (state: State) => state.value,
            (state: State, props: Props) => props.prop,
            jest.fn(() => ({})),
        )((state, props) => props.prop, {
            cacheObject: new CounterObjectCache()
        });

        const TestComponent: FunctionComponent<Props> = jest.fn(() => 'TestComponent Content');
        TestComponent.displayName = 'TestComponent';
        /*const ConnectedTestComponent = compose(
            connect(selector),
            reselectConnect(selector),
        )(TestComponent);*/
        const ConnectedTestComponent =
            connect(selector, {})(
                reselectConnect(selector)(
                    TestComponent
                )
            );


        const store = createStore(() => ({}), {});

        const wrapper = mount(
            <Provider store={store}>
                <ConnectedTestComponent prop="prop1"/>
            </Provider>
        ).children();

        /*wrapper.setProps({
            children: <ConnectedTestComponent prop="prop1"/>
        });

        wrapper.setProps({
            children: <ConnectedTestComponent prop="prop2"/>
        });

        wrapper.setProps({
            children: <ConnectedTestComponent prop="prop1"/>
        });*/

        expect(wrapper).toMatchSnapshot()
    });

    test('should clear cache after component unmount', () => {

    })
});
