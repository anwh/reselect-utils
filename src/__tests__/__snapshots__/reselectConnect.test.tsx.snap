// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reselectConnect should proxy all props and renders wrapped component 1`] = `
<Connect(InnerConnect(ReselectConnect(TestComponent)))
  prop="prop1"
>
  <InnerConnect(ReselectConnect(TestComponent))
    prop="prop1"
  >
    <ReselectConnect(TestComponent)
      prop="prop1"
    >
      <TestComponent
        prop="prop1"
      >
        TestComponent Content
      </TestComponent>
    </ReselectConnect(TestComponent)>
  </InnerConnect(ReselectConnect(TestComponent))>
</Connect(InnerConnect(ReselectConnect(TestComponent)))>
`;
